tests/test_timing_model.py

```diff
@@ -385,6 +387,56 @@ def test_parfile_and_timfile_jumps(timfile_jumps):
     assert fs[4] == "1"
 
 
+@pytest.mark.parametrize("sortkey", ["freq", "mjd_float"])
+def test_resorting_toas_residuals_match(timfile_nojumps, sortkey):
+    m = get_model(os.path.join(datadir, "NGC6440E.par"))
+    t = timfile_nojumps
+    r = pint.residuals.Residuals(t, m, subtract_mean=False)
+    tcopy = deepcopy(t)
+    i = np.argsort(t.table[sortkey])
+    tcopy.table = tcopy.table[i]
+    rsort = pint.residuals.Residuals(tcopy, m, subtract_mean=False)
+    assert np.all(r.time_resids[i] == rsort.time_resids)
+
+
+@pytest.mark.parametrize("sortkey", ["freq", "mjd_float"])
+def test_resorting_toas_chi2_match(timfile_nojumps, sortkey):
+    m = get_model(os.path.join(datadir, "NGC6440E.par"))
+    t = timfile_nojumps
+    r = pint.residuals.Residuals(t, m, subtract_mean=False)
+    tcopy = deepcopy(t)
+    i = np.argsort(t.table[sortkey])
+    tcopy.table = tcopy.table[i]
+    rsort = pint.residuals.Residuals(tcopy, m, subtract_mean=False)
+    # the differences seem to be related to floating point math
+    assert np.isclose(r.calc_chi2(), rsort.calc_chi2(), atol=1e-14)
+
+
+def test_shuffle_toas_residuals_match(timfile_nojumps):
+    m = get_model(os.path.join(datadir, "NGC6440E.par"))
+    t = timfile_nojumps
+    r = pint.residuals.Residuals(t, m, subtract_mean=False)
+    tcopy = deepcopy(t)
+    i = np.arange(len(t))
+    np.random.shuffle(i)
```

Comment:
Random shuffles are liable to cause a problem with tests that fail once and then not the next time. You can use `hypothesis`: `@given(permutations(np.arange(n)))`. This explores the space of permutations and finds the simplest one that fails the test.

Reply:
ok, good to know.  but I can't quite figure out the syntax to define n properly using a fixture or class-based method.  I just pushed up revised tests done in a single class but I had to define n using a global.  is there a better way?  when I tried to define it as a fixture there were various syntax errors.

Reply:
The magic sauce is that `n` is supposed to be the length of the TOAs object, so you need to make or pick the TOAs object within hypothesis. I think `flatmap` is the right way to do this: https://hypothesis.readthedocs.io/en/latest/data.html#chaining-strategies-together bit I'd find it easier to use `@composite`:

```python
@composite
def toas_and_order(draw):
    toas_name = draw(select_from(["file1.tim", "file2.tim"]))
    toas = get_TOAs(toas_name)
    n = len(toas)
    ix = draw(permutations(np.arange(n)))
    return toas, ix
```
then you'd use `@given(toas_and_order())` to get that tuple inside your function.

I think there are examples of `@composite` doing this sort of thing elsewhere in our test suite.

Reply:
That is handy.  I reworked it a bit so that it loads the data initially through a data class (unofficial) to speed things up, but I think this now does what it should